File: targets/java/javalin/src/main/java/rb/javalin/routes/Query.java
package rb.javalin.routes;

import io.javalin.config.JavalinConfig;
import io.javalin.http.Context;
import rb.domain.Domain;
import rb.domain.Model.QueryMany;
import rb.domain.Model.QueryOne;

/**
 * query: query string parsing, percent-decoding and coercion, with the values echoed beside the
 * small payload and put to no other use.
 *
 * Javalin's own binding: queryParamAsClass names the parameter and the class it wants, and
 * returns a Validator that has already run Javalin's converter for that type. getOrDefault is
 * what an absent one is; a value the converter refuses is a ValidationException, which is
 * Javalin's own 400 and not this repository's.
 */
public final class Query {
  private Query() {}

  // rb:wiring query.*,domain.*
  static int qint(Context ctx, String name, int fallback) {
    return ctx.queryParamAsClass(name, Integer.class).getOrDefault(fallback);
  }

  // rb:wiring query.*,domain.*
  static String qstr(Context ctx, String name) {
    return ctx.queryParamAsClass(name, String.class).getOrDefault("");
  }

  public static void register(JavalinConfig cfg) {
    cfg.routes.get("/query/one",
                   ctx -> ctx.json(Domain.withEcho("small", new QueryOne(qint(ctx, "page", 0)))));

    cfg.routes.get("/query/many", ctx -> ctx.json(Domain.withEcho("small", new QueryMany(
        qint(ctx, "page", 0), qint(ctx, "size", 0), qstr(ctx, "status"),
        qstr(ctx, "category"), qstr(ctx, "sort"), qstr(ctx, "q"),
        qint(ctx, "min_price", 0), qint(ctx, "max_price", 0)))));
  }
}

File: targets/node/fastify/query.js
// Fastify's own query binding: a JSON Schema on the route, compiled and run by the ajv it
// already bundles.
//
// Declaring `schema.querystring` is the whole wiring. ajv runs with coerceTypes: 'array'
// and removeAdditional: true, which are Fastify's own defaults, so a declared integer
// arrives as a number and anything undeclared is dropped before the handler sees it. The
// handler echoes what the schema left in req.query beside the small payload.
//
// No `required`, so Fastify decides what a missing parameter is: the property is simply
// absent. A value it cannot coerce is FST_ERR_VALIDATION, Fastify's own envelope, not this
// repository's. The endpoint set sends neither.

const page = { type: "integer" };

// rb:wiring query.*
/** query.one: one parameter, and nothing else survives the schema. */
export const bindsOne = {
  schema: { querystring: { type: "object", properties: { page } } },
};

// rb:wiring query.*
/** query.many: eight parameters, mixed types. */
export const bindsMany = {
  schema: {
    querystring: {
      type: "object",
      properties: {
        page,
        size: { type: "integer" },
        status: { type: "string" },
        category: { type: "string" },
        sort: { type: "string" },
        q: { type: "string" },
        min_price: { type: "integer" },
        max_price: { type: "integer" },
      },
    },
  },
};

// rb:wiring domain.*
/** What domain.filter pages by. */
export const bindsFilter = {
  schema: {
    querystring: {
      type: "object",
      properties: { page, size: { type: "integer" }, status: { type: "string" } },
    },
  },
};
